public/app/features/alerting/unified/utils/notification-policies.ts

```diff
@@ -1,28 +1,121 @@
+import {
+  AlertmanagerGroup,
+  MatcherOperator,
+  ObjectMatcher,
+  Route,
+  RouteWithID,
+} from 'app/plugins/datasource/alertmanager/types';
+import { Labels } from 'app/types/unified-alerting-dto';
+
+import { Label, normalizeMatchers } from './matchers';
+
+type OperatorPredicate = (labelValue: string, matcherValue: string) => boolean;
+
+const OperatorFunctions: Record<MatcherOperator, OperatorPredicate> = {
+  [MatcherOperator.equal]: (lv, mv) => lv === mv,
+  [MatcherOperator.notEqual]: (lv, mv) => lv !== mv,
+  [MatcherOperator.regex]: (lv, mv) => Boolean(lv.match(new RegExp(mv))),
+  [MatcherOperator.notRegex]: (lv, mv) => !Boolean(lv.match(new RegExp(mv))),
+};
+
+function isLabelMatch(matcher: ObjectMatcher, label: Label) {
+  const [labelKey, labelValue] = label;
+  const [matcherKey, operator, matcherValue] = matcher;
+
+  // not interested, keys don't match
+  if (labelKey !== matcherKey) {
+    return false;
+  }
+
+  const matchFunction = OperatorFunctions[operator];
+  if (!matchFunction) {
+    throw new Error(`no such operator: ${operator}`);
+  }
+
+  return matchFunction(labelValue, matcherValue);
+}
+
+interface LabelMatchResult {
+  match: boolean;
+  matchers: ObjectMatcher[];
+}
+
+interface MatchingResult {
+  matches: boolean;
+  details: Map<ObjectMatcher, Label[]>;
+  labelsMatch: Map<Label, LabelMatchResult>;
+}
+
+// check if every matcher returns "true" for the set of labels
+function matchLabels(matchers: ObjectMatcher[], labels: Label[]): MatchingResult {
+  const details = new Map<ObjectMatcher, Label[]>();
+
+  // If a policy has no matchers it still can be a match, hence matchers can be empty and match can be true
+  // So we cannot use empty array of matchers as an indicator of no match
+  const labelsMatch = new Map<Label, { match: boolean; matchers: ObjectMatcher[] }>(
+    labels.map((label) => [label, { match: false, matchers: [] }])
+  );
+
+  const matches = matchers.every((matcher) => {
+    const matchingLabels = labels.filter((label) => isLabelMatch(matcher, label));
+
+    matchingLabels.forEach((label) => {
+      const labelMatch = labelsMatch.get(label);
+      // The condition is just to satisfy TS. The map should have all the labels due to the previous map initialization
+      if (labelMatch) {
+        labelMatch.match = true;
+        labelMatch.matchers.push(matcher);
+      }
+    });
 
-import { Label, normalizeMatchers, labelsMatchObjectMatchers } from './matchers';
+    if (matchingLabels.length === 0) {
+      return false;
+    }
+
+    details.set(matcher, matchingLabels);
+    return matchingLabels.length > 0;
+  });
+
+  return { matches, details, labelsMatch };
+}
+
+export interface AlertInstanceMatch {
+  instance: Labels;
+  matchDetails: Map<ObjectMatcher, Labels>;
+  labelsMatch: Map<Label, LabelMatchResult>;
+}
+
+export interface RouteMatchResult<T extends Route> {
+  route: T;
+  details: Map<ObjectMatcher, Label[]>;
+  labelsMatch: Map<Label, LabelMatchResult>;
```

Comment:
TODO `details` is much more convenient to use. let's try to remove `labelsMatch`

Reply:
we are going to do this in a follow up PR